diff --git a/dlls/dsound/tests/ds3d.c b/dlls/dsound/tests/ds3d.c
index e8c2e32..ce049df 100644
--- a/dlls/dsound/tests/ds3d.c
+++ b/dlls/dsound/tests/ds3d.c
@@ -464,6 +464,7 @@ void test_buffer(LPDIRECTSOUND dso, LPDIRECTSOUNDBUFFER *dsbo,
         }
 
         /* Check for primary buffer size change */
+        todo_wine
         ok(new_dsbcaps.dwBufferBytes == dsbcaps.dwBufferBytes,
            "    buffer size changed after SetFormat() - "
            "previous size was %u, current size is %u\n",
diff --git a/dlls/dsound/tests/ds3d8.c b/dlls/dsound/tests/ds3d8.c
index 41de792..e316f14 100644
--- a/dlls/dsound/tests/ds3d8.c
+++ b/dlls/dsound/tests/ds3d8.c
@@ -305,6 +305,7 @@ void test_buffer8(LPDIRECTSOUND8 dso, LPDIRECTSOUNDBUFFER * dsbo,
         }
 
         /* Check for primary buffer size change */
+        todo_wine
         ok(new_dsbcaps.dwBufferBytes == dsbcaps.dwBufferBytes,
            "    buffer size changed after SetFormat() - "
            "previous size was %u, current size is %u\n",
