diff --git a/dlls/oleaut32/variant.c b/dlls/oleaut32/variant.c
index 0d1f2cc..95b5845 100644
--- a/dlls/oleaut32/variant.c
+++ b/dlls/oleaut32/variant.c
@@ -2086,7 +2086,7 @@ HRESULT WINAPI VarNumFromParseNum(NUMPARSE *pNumprs, BYTE *rgbDig,
       V_UI4(pVarDst) = ul64;
       return S_OK;
     }
-    else if (dwVtBits & VTBIT_I8 && ((ul64 <= I8_MAX)||(l64>=I8_MIN)))
+    else if (dwVtBits & VTBIT_I8)
     {
       V_VT(pVarDst) = VT_I8;
       V_I8(pVarDst) = ul64;
