diff --git a/tools/runtest b/tools/runtest
index a3c9331..ffbdf36 100755
--- a/tools/runtest
+++ b/tools/runtest
@@ -47,6 +47,8 @@ EOF
 platform=$WINETEST_PLATFORM
 WINETEST_DEBUG=${WINETEST_DEBUG:-1}
 
+cmdline="$0 $*"
+
 # parse command-line options
 while [ "$#" != 0 ]; do
     case "$1" in
@@ -124,4 +126,17 @@ export WINETEST_PLATFORM WINETEST_DEBUG
 #  WINETEST_WRAPPER=valgrind
 # would run the tests under valgrind to look for memory errors.
 
-exec $WINETEST_WRAPPER "$topobjdir/wine" "$program" "$infile" "$@"
+# If user ran "make -j2" or the like, use a temporary log file to
+# prevent mushing results from different tests together
+case x"$MAKEFLAGS"x in
+*-j*) 
+  logfile=`basename "$infile" .c`.log
+  echo "# $cmdline" > $logfile
+  $WINETEST_WRAPPER "$topobjdir/wine" "$program" "$infile" "$@" >> $logfile 2>&1
+  RETVAL=$?
+  cat $logfile
+  rm $logfile
+  exit $RETVAL ;;
+*) 
+  exec $WINETEST_WRAPPER "$topobjdir/wine" "$program" "$infile" "$@" ;;
+esac
