--- depot_tools/bootstrap/win/win_tools.bat.old	2010-01-16 21:43:10.000000000 -0800
+++ depot_tools/bootstrap/win/win_tools.bat	2010-01-16 21:43:43.000000000 -0800
@@ -1,4 +1,4 @@
-@echo off
+@echo on
 :: Copyright (c) 2009 The Chromium Authors. All rights reserved.
 :: Use of this source code is governed by a BSD-style license that can be
 :: found in the LICENSE file.
@@ -34,20 +34,28 @@
 echo Installing subversion ...
 :: svn is not accessible; check it out and create 'proxy' files.
 if exist "%~dp0svn.zip" del "%~dp0svn.zip"
-cscript //nologo //e:jscript "%~dp0get_file.js" %WIN_TOOLS_ROOT_URL%/third_party/svn_bin.zip "%~dp0svn.zip"
+
+rem cscript //nologo //e:jscript "%~dp0get_file.js" %WIN_TOOLS_ROOT_URL%/third_party/svn_bin.zip "%~dp0svn.zip"
+c:\cygwin\bin\wget %WIN_TOOLS_ROOT_URL%/third_party/svn_bin.zip
+move svn_bin.zip "%~dp0svn.zip"
+
 if errorlevel 1 goto :SVN_FAIL
 :: Cleanup svn directory if it was existing.
 if exist "%WIN_TOOLS_ROOT_DIR%\svn\." rd /q /s "%WIN_TOOLS_ROOT_DIR%\svn"
 if exist "%WIN_TOOLS_ROOT_DIR%\svn_bin\." rd /q /s "%WIN_TOOLS_ROOT_DIR%\svn_bin"
 :: Will create svn_bin\...
-cscript //nologo //e:jscript "%~dp0unzip.js" "%~dp0svn.zip" "%WIN_TOOLS_ROOT_DIR%"
+
+rem Work around cscript not working on wine
+rem cscript //nologo //e:jscript "%~dp0unzip.js" "%~dp0svn.zip" "%WIN_TOOLS_ROOT_DIR%"
+c:\cygwin\bin\unzip "%~dp0svn.zip" -d depot_tools
+
 if errorlevel 1 goto :SVN_FAIL
 if not exist "%WIN_TOOLS_ROOT_DIR%\svn_bin\." goto :SVN_FAIL
 del "%~dp0svn.zip"
 :: Create the batch file.
-call copy /y "%~dp0svn.new.bat" "%WIN_TOOLS_ROOT_DIR%\svn.bat" 1>nul
-call copy /y "%~dp0svnversion.new.bat" "%WIN_TOOLS_ROOT_DIR%\svnversion.bat" 1>nul
-call copy /y "%~dp0svn.new" "%WIN_TOOLS_ROOT_DIR%\svn" 1>nul
+copy /y "%~dp0svn.new.bat" "%WIN_TOOLS_ROOT_DIR%\svn.bat" 1>nul
+copy /y "%~dp0svnversion.new.bat" "%WIN_TOOLS_ROOT_DIR%\svnversion.bat" 1>nul
+copy /y "%~dp0svn.new" "%WIN_TOOLS_ROOT_DIR%\svn" 1>nul
 goto :PYTHON_CHECK
 
 
@@ -78,11 +86,11 @@
 :: Cleanup python directory if it was existing.
 if exist "%WIN_TOOLS_ROOT_DIR%\python\." rd /q /s "%WIN_TOOLS_ROOT_DIR%\python"
 if exist "%WIN_TOOLS_ROOT_DIR%\python_bin\." rd /q /s "%WIN_TOOLS_ROOT_DIR%\python_bin"
-call svn co -q %WIN_TOOLS_ROOT_URL%/third_party/python "%WIN_TOOLS_ROOT_DIR%\python_bin"
+call svn.bat co -q %WIN_TOOLS_ROOT_URL%/third_party/python "%WIN_TOOLS_ROOT_DIR%\python_bin"
 if errorlevel 1 goto :PYTHON_FAIL
 :: Create the batch file.
-call copy /y "%~dp0python.new.bat" "%WIN_TOOLS_ROOT_DIR%\python.bat" 1>nul
-call copy /y "%~dp0python.new" "%WIN_TOOLS_ROOT_DIR%\python" 1>nul
+copy /y "%~dp0python.new.bat" "%WIN_TOOLS_ROOT_DIR%\python.bat" 1>nul
+copy /y "%~dp0python.new" "%WIN_TOOLS_ROOT_DIR%\python" 1>nul
 set ERRORLEVEL=0
 goto :END
 
