diff --git a/tools/winewrapper b/tools/winewrapper
index b324daa..a96bc5b 100755
--- a/tools/winewrapper
+++ b/tools/winewrapper
@@ -88,6 +89,8 @@ if [ -x "$topdir/loader/$name" ]
 then WINELOADER="$topdir/loader/$name"
 elif [ -x "$topdir/loader/wine" ]
 then WINELOADER="$topdir/loader/wine"
+elif [ -x "$topdir/loader/wine64" ]
+then WINELOADER="$topdir/loader/wine64"
 else
   echo "$name: could not find the Wine loader in $topdir"
   exit 1
